### PR TITLE
update use of autotools

### PR DESCRIPTION
**Pull request by kerolasa**
There is quite a lot going on in this change.  The sort summary is: this is
a build system modernization.  Needless to say there is no way to do this
kind of update without causing some compatibility issues, so lets hope they
do not bother users too much.  Various changes includes:

* Generated files are removed from version control.  When building from git
  clone one has to first run ./autogen.sh
* Use pkg-config when ever possible.
* Rename command 'convert' to 'conserver.convert' to avoid collision with
  a command of same name from imagemagick.
* --with-logfile=PATH and --with-pidfile=PATH defaults include prefix.
* 'make test' is renamed to 'make check'.  Later is part of autotools
  standard invocation.
* Add --with-chat and --with-autologin options to build these command.  Both
  are disabled by default.  When enabled these commands are added to
  installation with manual pages.
* Tests work when performing out-of-tree build.  That said distribution
  packages should be created with 'make distcheck'.
* Distribution package is compressed with 'xz', and 'gzip' is omitted.
* The script build-aux/git-version-gen is copied from gnulib to this
  project.  Aim is to make version numbers to be derived from git tag name,
  but this change will need to wait later.
* Add .gitignore file.

All changes to source files are due technical necessity.  These sorts of
changes are kept absolutely minimal in this change, that aims to be as pure
build system update as possible.

But there are few changes, such as in chat.c function logf() had to be
renamed because glibc provides a logarithmic function of same name.  In
autologin.c setpgrp() used arguments, that is rather strange because even
posix defines this function to be used without them.  And project specific
include files are defined quotes rather than angle brackets to give visual
hint what is coming from where.

p.s. (not in git commit) Please don't worry if you autotools are not your
strong point. I can promise to help with build system, when needed.

### CODE REVIEW

**Comment by conserver** [comment]
Thanks for all this!

I haven't had a chance to go through it all, but there is one thing that jumps out - the removal of generated files.  I like the idea of getting them out of source control (before moving to git, there weren’t actually under source control, but in trying to replicate the old tar balls, I put them in so history could be recreated).  They do need to  be in the final release so folks can download the bundle and only need a compiler vs a full autotools environment - but maybe that happens with what you put together.  Again, I need to play around and see what’s what.

I don't see any major compatibility issues either - we'll just bump the version, call out the changes, and roll forward.

**Comment by kerolasa** [comment]
Good to hear this contribution is being reviewed. I've done some build-tools contributions to couple projects, and reactions vary. Sounds like conserver is open to get standard build working, that will make distribution packaging and stuff like that very trivial.
What comes to downloads 'make dist' will create a distribution file, that has various autotools generated files. That is the idea of distribution tar balls. Version control ought not to have these generated files. Developers ought to use their own copy of autotools to get things going. Yes, that is a bit of a ask to folks who want to do development but nothing extra ordinary - most of projects do exactly that.
Please do not bump version yet. Assuming the first merge goes through I'll make couple additional changes to ease maintenance work. Trust me, releasing can be made very easy indeed.

**Comment by kerolasa** [comment]
FYI couple more commits added to this pull, that implement missing pieces I mentioned yesterday. I also noticed ./configure end summary went missing in first try and now it's back.

So assuming merged then a version update is a matter of adding a 'git tag -a vX.Y.Z' followed by clean build after 'git clean -xdf' and 'make distcheck'. The .tar.xz will need gpg signing and uploading, both can be quite trivially scripted and I leave that to you as no-one else is ever going to need the script.

**Comment by kerolasa** [comment]
Original pull did not check openpty and setproctitle function correctly. That's now fixed.
While waiting this to be merged I've got next pull ready, and waitin in my repo. Third thing is nearly done, that introduces json output that I am planning to use when monitoring conserver. But these need to wait until the build system pull (this one) is merged.

**Comment by Lyude** [comment]
Hi, could this get merged? Came here because I wanted to add some small features to conserver, but it's rather annoying to have git keep pestering me just by re-running autoreconf :s

**Comment by Lyude** [comment]
...ah. now that I look at this closer I can see why this isn't merged. I might as well go leave some comments on the broken commits here :)

File: configure.ac [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  sysconf \
+  tcgetattr \
+  tcgetpgrp \
+  tcsendbreak \
+  tcsetattr \
+  tcsetpgrp \
+  unlockpt \
+])
+
+AC_SEARCH_LIBS([openpty], [util], [UTIL_LIBS=""], [UTIL_LIBS="-lutil"])
```
Comment (Lyude) [inline configure.ac]: You got this backwards! This should be:

```suggestion
AC_SEARCH_LIBS([openpty], [util], [UTIL_LIBS="-lutil"], [UTIL_LIBS=""])
```

File: configure.ac [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+AS_IF([test "x$ac_cv_search_openpty" != "xno"],
+  [AC_DEFINE([HAVE_OPENPTY], [1], [Defined if openpty exists])])
+AC_SUBST([UTIL_LIBS])
+
+AC_SEARCH_LIBS([setproctitle], [bsd], [BSD_LIBS=""], [BSD_LIBS="-lbsd"])
```
Comment (Lyude) [inline configure.ac]: Same for this one:

```suggestion
AC_SEARCH_LIBS([setproctitle], [bsd], [BSD_LIBS="-lbsd"], [BSD_LIBS=""])
```

**Comment by kerolasa** [comment]
Thank you for review @Lyude. Commit 262629209a4a62eeb33d33822b9cf0bbbf8790d8 updated comment near openpty() check and while on it I notice libbsd has started providing pkg-config support so that is now in use.

As mentioned by @Lyude it would be nice to autotools stuff fully reviewed, and merged to upstream so that https://github.com/conserver/conserver/pull/4 becomes unblocked. And I have couple changes idling in my clone that has not even reached review yet.

**Comment by conserver** [comment]
The activity here got me to start looking at this again.  I kinda put it aside because of comments in #4 about multi-platform / backward compatibility.  I'm attempting to take this PR and adjust things so that the code compiles and runs locally on my mac, on things like centos / ubuntu / etc and then hope that things like cross-compiling for other systems still work.  The changes here work on the environment they were built on (from what I remember), but break across other systems.  That's something I don't want to see happen.

I'm slowly slogging through the various issues and trying to undo some things so that the compatibility layers all work out.  Dunno if I'll submit a PR to @kerolasa's fork or just suck it all in, but as long as I don't get too distracted, I'll keep whacking on things.  Then #4 will be up for review/fixes.

Last time I did this there were a multitude of things and I got to the point of just giving up.  Fingers crossed with this round.  :-/

**Comment by Lyude** [comment]
> The activity here got me to start looking at this again. I kinda put it aside because of comments in #4 about multi-platform / backward compatibility. I'm attempting to take this PR and adjust things so that the code compiles and runs locally on my mac, on things like centos / ubuntu / etc and then hope that things like cross-compiling for other systems still work. The changes here work on the environment they were built on (from what I remember), but break across other systems. That's something I don't want to see happen.
> 
> I'm slowly slogging through the various issues and trying to undo some things so that the compatibility layers all work out. Dunno if I'll submit a PR to @kerolasa's fork or just suck it all in, but as long as I don't get too distracted, I'll keep whacking on things. Then #4 will be up for review/fixes.
> 
> Last time I did this there were a multitude of things and I got to the point of just giving up. Fingers crossed with this round. :-/

As a note: have you guys considered eventually moving over to meson from autotools? Would probably make all of this a lot easier

**Comment by conserver** [comment]
> As a note: have you guys considered eventually moving over to meson from autotools? Would probably make all of this a lot easier

Nope - mainly because I didn't even know it existed, but also because the autotools worlds are known to many.  Doesn't mean it shouldn't/couldn't happen eventually, but certainly not on my radar.

**Comment by conserver** [comment]
So, I keep trying to unravel all the changes here that break things and instrument a fix.  I've found things like missing checks for header files in configure.ac, extra includes in .h files that should come from compat.h, broken support for unix domain sockets, missing library checks, and a broken test system.  I believe there are also code changes that break systems I haven't even tried to test (never got past my Mac, honestly).  All this, and I haven't even gotten though the changes - just been working from the autotools layer down and that's highlighted some of the lower-level stuff.

On the flip side, there are a lot of cool things in here.  I wish I could easily pick and choose the parts that don't cause systems to break.

A bit on my perspective: This software has been used in various forms on a variety of platforms over the years, from being embedded in console servers to deployed across large fleets of servers using both linux and non-linux based environments (and even various generations and distros of linux).  I can't just turn my back on all those possibilities when the real desire here is to adopt more of the autotools suite.  If the autotools portions could be done without requiring the "latest" linux with the "latest" set of tools, then great (quotes around that 'cause it's a fuzzy statement).  I do believe Linux is/has been taking over the world, but I do know that there are shops running multi-year old distributions (not for the better) that shouldn't get left out.

Bottom line, I'm battling too many issues here to take this as is.  I suggest that either additional changes are made to help fix the issues (mainly from configure.ac missing things that configure.in did, I think), or perhaps smaller, bite-sized changes come in to add features without breaking the compatibility.  Not sure what the right answer is, but I've attempted to fix the issues a couple times now and each time the more I peeled back the layers, the more I found that wasn't working, and I just couldn't slog through it.

**Comment by kerolasa** [comment]
Conflicts resolved. Lets see if I can borrow a mac and make it compile. Are there other platforms that should be made to work, and would it be possible to provide access to such if they are important.

>  If the autotools portions could be done without requiring the "latest" linux with the "latest" set of tools

Ehm, these changes require autoconf 2.59 (released November 2003) and automake 1.7.9 (also Nov 2003). I strongly disagree these being unreasonable requirements.

p.s. Recent push changed upstream master history. Please avoid force pushing changes to upstream origin/master, such causes quite confusing situation to contributors.

**Comment by conserver** [comment]
> Ehm, these changes require autoconf 2.59 (released November 2003) and automake 1.7.9 (also Nov 2003). I strongly disagree these being unreasonable requirements.

The autotools bits aren't the problem - didn't mean to imply that with my blanket statement.  I was always trying to use a fairly recent version of things to build the configure script before publishing a new version.  One example was the change to setpgrp() - classic BSD wants arguments (see https://www.freebsd.org/cgi/man.cgi?sektion=2&query=setpgrp).  I was lumping that into "abandoning all but the latest linux" in my head.  :-/

**Comment by robohack** [comment]
> > As a note: have you guys considered eventually moving over to meson from autotools? Would probably make all of this a lot easier
> 
> Nope - mainly because I didn't even know it existed, but also because the autotools worlds are known to many. Doesn't mean it shouldn't/couldn't happen eventually, but certainly not on my radar.

Meson is not much better than CMake, and CMake is an absolute disaster of a monster.  The main difference is Python vs. C++, but neither are anywhere near as nuanced as Autoconf + Automake, and neither really actually do what they claim (at least in 99.99% of actual uses) -- the end user still needs them, whereas Autotools actually achieves that goal and the end user only needs GMake (plus the ultimate toolchain bits for compiling and linking, of course).

A better solution than all of those so-called configuration and build tools would be a good code cleanup and some simple BSD makefiles with possibly a very few very simple system dependent configuration templates where pure ifdefs won't work.  Then non-BSD users would have to install BSD Make (aka BMake), but that's easy.  :-)
